File: 28.实现-str-str.go
/*
 * @lc app=leetcode.cn id=28 lang=golang
 *
 * [28] 实现 strStr()
 */
package main

import "fmt"

// @lc code=start
func strStr(haystack string, needle string) int {
	haystackLen := len(haystack)
	needleLen := len(needle)
	res := -1
	for i := 0; i < haystackLen-needleLen+1; i++ {
		if haystack[i] == needle[0] && haystack[i:i+needleLen] == needle {
			fmt.Println(haystack[i : i+needleLen])
			return i
		}
	}
	return res
}

// @lc code=end

File: 28.实现-str-str.rs
/*
 * @lc app=leetcode.cn id=28 lang=rust
 *
 * [28] 实现 strStr()
 */

// @lc code=start
impl Solution {
    pub fn str_str(haystack: String, needle: String) -> i32 {
        let haystack: Vec<char> = haystack.chars().collect();
        let needle: Vec<char> = needle.chars().collect();
        let haystack_len = haystack.len();
        let needle_len = needle.len();
        if haystack_len < needle_len {
            return -1;
        }
        for i in 0..=haystack_len - needle_len {
            if haystack[i] == needle[0] && &haystack[i..(i + needle_len)] == &needle {
                return i as i32;
            }
        }
        -1
    }
}
// @lc code=end

